src/H5TSint.c

```diff
@@ -556,18 +568,30 @@ H5TS__tinfo_term(void)
         HGOTO_DONE(FAIL);
     while (H5TS_tinfo_next_free_s) {
         H5TS_tinfo_node_t *next = H5TS_tinfo_next_free_s->next;
+
+        /* Free any entries within the error stack */
+        H5E__clear_stack(&H5TS_tinfo_next_free_s->info.err_stack);
+
         H5MM_free(H5TS_tinfo_next_free_s);
         H5TS_tinfo_next_free_s = next;
     }
-    if (H5_UNLIKELY(H5TS_mutex_unlock(&H5TS_tinfo_mtx_s) < 0))
+
+    /* Release the thread info node for this thread since it's not on the free list */
+    if (H5_UNLIKELY(H5TS_key_get_value(H5TS_thrd_info_key_g, (void **)&tinfo_node) < 0))
         HGOTO_DONE(FAIL);
 
-    /* Release critical section / mutex for modifying the thread info globals */
-    if (H5_UNLIKELY(H5TS_mutex_destroy(&H5TS_tinfo_mtx_s) < 0))
+    if (tinfo_node) {
+        /* Free any entries within the error stack */
+        H5E__clear_stack(&tinfo_node->info.err_stack);
+        H5MM_free(tinfo_node);
+    }
+
+    /* Clear the thread info key to prevent bad reuse if the library is re-opened later in this process */
+    if (H5_UNLIKELY(H5TS_key_set_value(H5TS_thrd_info_key_g, NULL) < 0))
```

Comment:
Have you seen evidence of this reuse?

Reply:
I saw it during testing, but I think this is another holdover from an initial design that had two H5TS term routines. Removed in newer commit.